File: scripts/plot_shower_hits.jl
#!/usr/bin/env julia
"""
Plot shower particle hits on the detector box and terrain mesh.

Two-panel figure in local ENU coordinates:

  Left   – 300 m top-down view: terrain triangles coloured by altitude,
            box footprint (red), particle hits coloured by log10(KE/GeV).

  Right  – Zoom ±15 m around box centre: box triangle edges (red),
            particle hits coloured by log10(KE/GeV).

Usage
-----
  julia --project=. scripts/plot_shower_hits.jl \\
      TERRAIN.h5 BOX.ply PARTICLES.parquet [options]

  Options:
    --output PATH    Output PNG file   [default: shower_hits.png]
    --radius M       Terrain radius in left panel (m)  [default: 300]
    --zoom M         Half-width of zoom panel (m)      [default: 15]
    --dpi N          Resolution                        [default: 200]
"""

using ArgParse
using CairoMakie
using HDF5
using LinearAlgebra
using Parquet2
using Statistics

const EARTH_RADIUS_M = 6.371e6

# ---------------------------------------------------------------------------
# Shared helpers (same as plot_terrain_box.jl)
# ---------------------------------------------------------------------------

function ecef_to_enu(ecef::AbstractMatrix, lat0_deg::Float64, lon0_deg::Float64)
    lat0 = deg2rad(lat0_deg)
    lon0 = deg2rad(lon0_deg)
    p0   = EARTH_RADIUS_M .* [cos(lat0)*cos(lon0), cos(lat0)*sin(lon0), sin(lat0)]
    east  = [-sin(lon0),  cos(lon0),  0.0]
    north = [-sin(lat0)*cos(lon0), -sin(lat0)*sin(lon0),  cos(lat0)]
    up    = [ cos(lat0)*cos(lon0),  cos(lat0)*sin(lon0),  sin(lat0)]
    R  = vcat(east', north', up')
    dp = ecef .- p0'
    return (R * dp')'
end

function _ply_type_reader(t::AbstractString)
    t in ("float",  "float32") && return io -> Float64(ltoh(read(io, Float32)))
    t in ("double", "float64") && return io -> ltoh(read(io, Float64))
    t in ("int",    "int32")   && return io -> Int(ltoh(read(io, Int32)))
    t in ("uint",   "uint32")  && return io -> Int(ltoh(read(io, UInt32)))
    t in ("uchar",  "uint8")   && return io -> Int(read(io, UInt8))
    t in ("short",  "int16")   && return io -> Int(ltoh(read(io, Int16)))
    error("Unsupported PLY type: $t")
end

function read_ply(filepath::AbstractString)
    open(filepath, "r") do io
        n_verts = 0; n_faces = 0
        vertex_props = Tuple{String,String}[]
        face_count_type = "uchar"; face_index_type = "uint"
        current = :none
        while true
            line = strip(readline(io))
            line == "end_header" && break
            tokens = split(line); isempty(tokens) && continue
            if tokens[1] == "element"
                current = Symbol(tokens[2])
                tokens[2] == "vertex" && (n_verts = parse(Int, tokens[3]))
                tokens[2] == "face"   && (n_faces = parse(Int, tokens[3]))
            elseif tokens[1] == "property" && current == :vertex && tokens[2] != "list"
                push!(vertex_props, (tokens[end], tokens[2]))
            elseif tokens[1] == "property" && current == :face && tokens[2] == "list"
                face_count_type = tokens[3]; face_index_type = tokens[4]
            end
        end
        xi = findfirst(p -> p[1] == "x", vertex_props)
        yi = findfirst(p -> p[1] == "y", vertex_props)
        zi = findfirst(p -> p[1] == "z", vertex_props)
        readers = [_ply_type_reader(p[2]) for p in vertex_props]
        verts = zeros(Float64, n_verts, 3)
        for i in 1:n_verts
            vals = [r(io) for r in readers]
            verts[i, 1] = vals[xi]; verts[i, 2] = vals[yi]; verts[i, 3] = vals[zi]
        end
        cnt_reader = _ply_type_reader(face_count_type)
        idx_reader = _ply_type_reader(face_index_type)
        faces = Vector{Vector{Int}}(undef, n_faces)
        for j in 1:n_faces
            cnt = cnt_reader(io)
            faces[j] = [idx_reader(io) for _ in 1:cnt]
        end
        return verts, faces
    end
end

# ---------------------------------------------------------------------------
# Parquet reader
# ---------------------------------------------------------------------------

"""
Read particle positions and kinetic energies from a CORSIKA 8 parquet file.

Returns (xyz_ecef, ke_gev, pdg) where xyz_ecef is (N×3) Float64 in metres,
ke_gev is (N,) Float64 in GeV, and pdg is (N,) Int64.
Prints available columns so the user can check.
"""
function read_particles(path::AbstractString)
    ds  = Parquet2.Dataset(path)
    tbl = Parquet2.read(ds)
    cols = propertynames(tbl)
    println("  Parquet columns: ", join(string.(cols), ", "))

    # Position: try "x","y","z" then "position_x" etc.
    function getcol(names...)
        for n in names
            sym = Symbol(n)
            sym in cols && return Vector{Float64}(getproperty(tbl, sym))
        end
        error("None of $(names) found in parquet columns: $(cols)")
    end

    x  = getcol("x", "position_x", "X")
    y  = getcol("y", "position_y", "Y")
    z  = getcol("z", "position_z", "Z")

    # Kinetic energy: try eV then GeV columns
    ke_raw = getcol("kinetic_energy", "energy", "ekin", "KineticEnergy")
    # Determine unit: if median > 1e6 it is likely in eV
    ke_gev = median(ke_raw) > 1e6 ? ke_raw ./ 1e9 : ke_raw

    # PDG code (optional)
    pdg_sym = findfirst(n -> n in cols, [Symbol("pdg_code"), Symbol("pdg"),
                                          Symbol("particle"), Symbol("code")])
    pdg = pdg_sym !== nothing ? Vector{Int}(getproperty(tbl, cols[pdg_sym])) :
                                fill(0, length(x))

    return hcat(x, y, z), ke_gev, pdg
end

# ---------------------------------------------------------------------------
# Terrain helpers
# ---------------------------------------------------------------------------

"""Filter terrain faces whose centroid ENU distance (E-N only) from origin < radius."""
function filter_terrain_faces(verts_enu::AbstractMatrix,
                               faces::AbstractMatrix,
                               radius::Float64)
    keep = Int[]
    for i in 1:size(faces, 1)
        idx = faces[i, :] .+ 1   # 0-based → 1-based
        ce  = mean(verts_enu[idx, 1])
        cn  = mean(verts_enu[idx, 2])
        sqrt(ce^2 + cn^2) <= radius && push!(keep, i)
    end
    return keep
end

# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------

function parse_args_main()
    s = ArgParseSettings(
        description = "Plot shower particle hits on terrain + box mesh.",
        add_help    = true,
    )
    @add_arg_table! s begin
        "terrain_h5"
            help = "Terrain geometry HDF5 (from create_terrain_geometry.jl)"
            arg_type = String; required = true
        "box_ply"
            help = "Box PLY file (from make_box_scene.jl)"
            arg_type = String; required = true
        "particles_parquet"
            help = "Particles parquet (from terrain_shower output/particles/)"
            arg_type = String; required = true
        "--output"
            help = "Output PNG file path"
            arg_type = String; default = "shower_hits.png"
        "--radius"
            help = "Terrain display radius in left panel (m)"
            arg_type = Float64; default = 300.0
        "--zoom"
            help = "Half-width of zoom panel around box (m)"
            arg_type = Float64; default = 15.0
        "--dpi"
            help = "Resolution (dots per inch)"
            arg_type = Int; default = 200
        "--group"
            help = "HDF5 group for terrain vertices/faces"
            arg_type = String; default = ""
    end
    return ArgParse.parse_args(s)
end

function find_default_group(fid::HDF5.File)::String
    for name in keys(fid)
        g = fid[name]
        isa(g, HDF5.Group) && haskey(g, "vertices") && haskey(g, "faces") && return name
    end
    error("No group with 'vertices' and 'faces' found.")
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

function main()
    args = parse_args_main()

    # ── Load terrain ──────────────────────────────────────────────────────────
    println("Reading terrain: $(args["terrain_h5"])")
    terrain_verts_ecef, terrain_faces_raw, lat0_deg, lon0_deg =
        h5open(args["terrain_h5"], "r") do f
            gname = isempty(args["group"]) ? find_default_group(f) : args["group"]
            println("  Group: $gname")
            g = f[gname]
            v  = Float64.(read(g["vertices"]))   # (N, 3)
            fa = Int.(read(g["faces"]))           # (M, 3), 0-based
            lat0 = Float64(read(HDF5.attributes(g)["lat_deg"]))
            lon0 = Float64(read(HDF5.attributes(g)["lon_deg"]))
            v, fa, lat0, lon0
        end
    println("  $(size(terrain_verts_ecef,1)) vertices, $(size(terrain_faces_raw,1)) faces")

    # ── Load box ──────────────────────────────────────────────────────────────
    println("Reading box: $(args["box_ply"])")
    box_verts_ecef, box_faces_raw = read_ply(args["box_ply"])
    println("  $(size(box_verts_ecef,1)) vertices, $(length(box_faces_raw)) faces")

    # ── Load particles ────────────────────────────────────────────────────────
    println("Reading particles: $(args["particles_parquet"])")
    part_xyz_ecef, part_ke_gev, part_pdg = read_particles(args["particles_parquet"])
    n_part = size(part_xyz_ecef, 1)
    println("  $(n_part) particle hits")
    if n_part == 0
        println("  No particles found — plot will show geometry only.")
    end

    # ── Project to ENU ────────────────────────────────────────────────────────
    terrain_enu  = ecef_to_enu(terrain_verts_ecef, lat0_deg, lon0_deg)
    box_enu      = ecef_to_enu(box_verts_ecef,     lat0_deg, lon0_deg)
    part_enu     = n_part > 0 ?
                   ecef_to_enu(part_xyz_ecef, lat0_deg, lon0_deg) :
                   zeros(0, 3)

    # ── Filter terrain faces within radius ────────────────────────────────────
    nearby = filter_terrain_faces(terrain_enu, terrain_faces_raw, args["radius"])
    println("  $(length(nearby)) terrain faces within $(args["radius"]) m")

    # ── Colour scale for kinetic energy ──────────────────────────────────────
    log_ke = n_part > 0 ? log10.(clamp.(part_ke_gev, 1e-4, Inf)) : Float64[]
    ke_lo  = n_part > 0 ? minimum(log_ke) : -2.0
    ke_hi  = n_part > 0 ? maximum(log_ke) :  4.0

    # Terrain altitude colour limits from visible faces.
    vis_verts = unique(vcat([terrain_faces_raw[i, :] .+ 1 for i in nearby]...))
    alt_lo = isempty(vis_verts) ? 0.0 : minimum(terrain_enu[vis_verts, 3])
    alt_hi = isempty(vis_verts) ? 1000.0 : maximum(terrain_enu[vis_verts, 3])

    # ── Figure ────────────────────────────────────────────────────────────────
    println("Building figure ...")
    set_theme!(theme_latexfonts())
    fig = Figure(size = (1100, 520))

    # ── Panel 1: 300 m top-down ───────────────────────────────────────────────
    ax1 = Axis(fig[1, 1],
               xlabel  = "East  (m)",
               ylabel  = "North  (m)",
               title   = "Top-down  $(round(Int, args["radius"])) m view",
               aspect  = DataAspect())

    # Terrain triangles coloured by mean altitude.
    for i in nearby
        idx  = terrain_faces_raw[i, :] .+ 1
        vx   = terrain_enu[idx, 1]
        vy   = terrain_enu[idx, 2]
        alt  = mean(terrain_enu[idx, 3])
        poly!(ax1, Point2f.(vx, vy);
              color       = alt,
              colormap    = :terrain,
              colorrange  = (alt_lo, alt_hi),
              strokecolor = (:black, 0.3),
              strokewidth = 0.5)
    end

    # Box footprint.
    e_lo, e_hi = extrema(box_enu[:, 1])
    n_lo, n_hi = extrema(box_enu[:, 2])
    box_rect = [Point2f(e_lo, n_lo), Point2f(e_hi, n_lo),
                Point2f(e_hi, n_hi), Point2f(e_lo, n_hi)]
    poly!(ax1, box_rect; color = (:red, 0.0),
          strokecolor = :red, strokewidth = 2)

    # Particle hits.
    sc1 = nothing
    if n_part > 0
        sc1 = scatter!(ax1, part_enu[:, 1], part_enu[:, 2];
                       color      = log_ke,
                       colormap   = :plasma,
                       colorrange = (ke_lo, ke_hi),
                       markersize = 8,
                       strokewidth = 0.5,
                       strokecolor = :black)
    end

    xlims!(ax1, -args["radius"], args["radius"])
    ylims!(ax1, -args["radius"], args["radius"])

    # Terrain colorbar.
    Colorbar(fig[1, 2]; colormap = :terrain, limits = (alt_lo, alt_hi),
             label = "Altitude  (m)", height = Relative(0.8))

    # ── Panel 2: zoom around box ──────────────────────────────────────────────
    z = args["zoom"]
    ax2 = Axis(fig[1, 3],
               xlabel  = "East  (m)",
               ylabel  = "North  (m)",
               title   = "Box zoom  ±$(round(Int,z)) m",
               aspect  = DataAspect())

    # Box triangle edges.
    for face in box_faces_raw
        idx = face .+ 1
        vx  = push!(box_enu[idx, 1], box_enu[idx[1], 1])
        vy  = push!(box_enu[idx, 2], box_enu[idx[1], 2])
        lines!(ax2, vx, vy; color = :red, linewidth = 0.8)
    end

    # Particle hits.
    sc2 = nothing
    if n_part > 0
        sc2 = scatter!(ax2, part_enu[:, 1], part_enu[:, 2];
                       color      = log_ke,
                       colormap   = :plasma,
                       colorrange = (ke_lo, ke_hi),
                       markersize = 10,
                       strokewidth = 0.8,
                       strokecolor = :black)
    end

    xlims!(ax2, -z, z)
    ylims!(ax2, -z, z)

    # Energy colorbar.
    if sc1 !== nothing || sc2 !== nothing
        Colorbar(fig[1, 4]; colormap = :plasma, limits = (ke_lo, ke_hi),
                 label = "log₁₀(KE / GeV)", height = Relative(0.8))
    end

    # Title.
    n_str = n_part > 0 ? "$(n_part) hits" : "no hits"
    Label(fig[0, 1:4],
          "Shower hits — lat=$(round(lat0_deg,digits=3))°  " *
          "lon=$(round(lon0_deg,digits=3))°  ($n_str)",
          fontsize = 13, tellwidth = false)

    # ── Save ─────────────────────────────────────────────────────────────────
    out_dir = dirname(abspath(args["output"]))
    isempty(out_dir) || mkpath(out_dir)
    save(args["output"], fig, px_per_unit = args["dpi"] / 72)
    println("Saved: $(args["output"])")
    println("Done.")
end

main()

File: scripts/plot_shower_hits.py
#!/usr/bin/env python3
"""
Plot shower particle hits on the detector box and terrain mesh.

Two-panel figure in local ENU coordinates:
  Left  – 300 m top-down view: terrain triangles coloured by altitude,
           box footprint (red), particle hits coloured by log10(KE/GeV).
  Right – Zoom ±15 m around box: box triangle edges (red),
           particle hits coloured by log10(KE/GeV).

Usage
-----
  python3 scripts/plot_shower_hits.py TERRAIN.h5 BOX.ply PARTICLES.parquet
      [--output PATH] [--radius M] [--zoom M] [--dpi N]
"""

import argparse
import struct
import sys

import h5py
import matplotlib.pyplot as plt
import matplotlib.tri as mtri
from matplotlib.patches import Polygon
from matplotlib.collections import PatchCollection
import numpy as np
import pyarrow.parquet as pq

EARTH_RADIUS_M = 6.371e6

# ---------------------------------------------------------------------------
# Coordinate helpers
# ---------------------------------------------------------------------------

def ecef_to_enu(ecef, lat0_deg, lon0_deg):
    """Convert (N,3) ECEF array to local ENU centred on (lat0, lon0), altitude=0."""
    lat0 = np.radians(lat0_deg)
    lon0 = np.radians(lon0_deg)
    p0 = EARTH_RADIUS_M * np.array([
        np.cos(lat0)*np.cos(lon0),
        np.cos(lat0)*np.sin(lon0),
        np.sin(lat0)])
    east  = np.array([-np.sin(lon0),  np.cos(lon0),  0.0])
    north = np.array([-np.sin(lat0)*np.cos(lon0), -np.sin(lat0)*np.sin(lon0), np.cos(lat0)])
    up    = np.array([ np.cos(lat0)*np.cos(lon0),  np.cos(lat0)*np.sin(lon0), np.sin(lat0)])
    R  = np.vstack([east, north, up])   # (3, 3)
    dp = ecef - p0                       # (N, 3)
    return (R @ dp.T).T                  # (N, 3) -> columns: East, North, Up

# ---------------------------------------------------------------------------
# PLY reader (binary_little_endian, double or float vertices)
# ---------------------------------------------------------------------------

def _ply_fmt(t):
    return {'float': ('f', 4), 'float32': ('f', 4),
            'double': ('d', 8), 'float64': ('d', 8),
            'int': ('i', 4), 'int32': ('i', 4),
            'uint': ('I', 4), 'uint32': ('I', 4),
            'uchar': ('B', 1), 'uint8': ('B', 1)}[t]

def read_ply(path):
    with open(path, 'rb') as f:
        n_verts = n_faces = 0
        vprops = []
        fcount_fmt = ('B', 1)
        fidx_fmt   = ('I', 4)
        current = None
        while True:
            line = f.readline().decode('ascii').strip()
            if line == 'end_header':
                break
            tok = line.split()
            if not tok:
                continue
            if tok[0] == 'element':
                current = tok[1]
                if tok[1] == 'vertex': n_verts = int(tok[2])
                if tok[1] == 'face':   n_faces = int(tok[2])
            elif tok[0] == 'property' and current == 'vertex' and tok[1] != 'list':
                vprops.append((tok[-1], tok[1]))
            elif tok[0] == 'property' and current == 'face' and tok[1] == 'list':
                fcount_fmt = _ply_fmt(tok[2])
                fidx_fmt   = _ply_fmt(tok[3])

        xi = next(i for i,(n,_) in enumerate(vprops) if n=='x')
        yi = next(i for i,(n,_) in enumerate(vprops) if n=='y')
        zi = next(i for i,(n,_) in enumerate(vprops) if n=='z')
        fmts = [_ply_fmt(t) for _,t in vprops]
        row_size = sum(s for _,s in fmts)

        verts = np.zeros((n_verts, 3))
        for i in range(n_verts):
            row = f.read(row_size)
            off = 0
            vals = []
            for fmt, size in fmts:
                vals.append(struct.unpack_from('<'+fmt, row, off)[0])
                off += size
            verts[i] = [vals[xi], vals[yi], vals[zi]]

        faces = []
        for _ in range(n_faces):
            cnt = struct.unpack('<'+fcount_fmt[0], f.read(fcount_fmt[1]))[0]
            idx = [struct.unpack('<'+fidx_fmt[0], f.read(fidx_fmt[1]))[0]
                   for _ in range(cnt)]
            faces.append(idx)

    return verts, faces

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('terrain_h5')
    ap.add_argument('box_ply')
    ap.add_argument('particles_parquet')
    ap.add_argument('--output', default='shower_hits.png')
    ap.add_argument('--radius', type=float, default=300.0,
                    help='Terrain display radius in left panel (m)')
    ap.add_argument('--zoom', type=float, default=15.0,
                    help='Half-width of zoom panel (m)')
    ap.add_argument('--dpi', type=int, default=200)
    ap.add_argument('--group', default='',
                    help='HDF5 group for terrain vertices/faces')
    args = ap.parse_args()

    # ── Load terrain ─────────────────────────────────────────────────────────
    print(f'Reading terrain: {args.terrain_h5}')
    with h5py.File(args.terrain_h5, 'r') as f:
        gname = args.group or next(
            k for k in f if 'vertices' in f[k] and 'faces' in f[k])
        print(f'  Group: {gname}')
        g = f[gname]
        # HDF5.jl stores (N,3) Julia arrays as (3,N) on disk; h5py reads (3,N)
        v_raw = g['vertices'][:]   # h5py shape: (3, N)
        fa    = g['faces'][:]      # h5py shape: (3, M), 0-based
        lat0  = float(g.attrs['lat_deg'])
        lon0  = float(g.attrs['lon_deg'])

    # h5py returns (3, N) -> transpose to (N, 3)
    terrain_ecef = v_raw.T if v_raw.shape[0] == 3 else v_raw
    terrain_faces = fa.T  if fa.shape[0]     == 3 else fa  # (M, 3)
    print(f'  {terrain_ecef.shape[0]} vertices, {terrain_faces.shape[0]} faces')

    # ── Load box ─────────────────────────────────────────────────────────────
    print(f'Reading box: {args.box_ply}')
    box_ecef, box_faces = read_ply(args.box_ply)
    print(f'  {len(box_ecef)} vertices, {len(box_faces)} faces')

    # ── Load particles ────────────────────────────────────────────────────────
    print(f'Reading particles: {args.particles_parquet}')
    tbl = pq.read_table(args.particles_parquet).to_pydict()
    px = np.array(tbl['x'], dtype=float)   # CORSIKA local x (≈ North, metres)
    py = np.array(tbl['y'], dtype=float)   # CORSIKA local y (≈ East,  metres)
    ke = np.array(tbl['kinetic_energy'], dtype=float)   # GeV
    pdg = np.array(tbl['pdg'], dtype=int)
    n_part = len(px)
    print(f'  {n_part} particle hits')
    pdg_labels = {22: 'γ', 11: 'e⁻', -11: 'e⁺', 13: 'μ⁻', -13: 'μ⁺'}
    for code, count in zip(*np.unique(pdg, return_counts=True)):
        print(f'    PDG {code:5d} ({pdg_labels.get(code,"?")}): {count}')

    # ── Project terrain & box to ENU ─────────────────────────────────────────
    terrain_enu = ecef_to_enu(terrain_ecef, lat0, lon0)  # (N,3): East, North, Up
    box_enu     = ecef_to_enu(box_ecef,     lat0, lon0)

    # Box centroid in ENU (should be ≈ (0, 0, altitude))
    box_cen = box_enu.mean(axis=0)
    print(f'  Box centroid ENU: East={box_cen[0]:.1f} m, '
          f'North={box_cen[1]:.1f} m, Alt={box_cen[2]:.1f} m')

    # Particle positions are ECEF offsets from the box centroid.
    # Convert to local ENU via the site rotation matrix.
    lat0_r = np.radians(lat0); lon0_r = np.radians(lon0)
    east_hat  = np.array([-np.sin(lon0_r),  np.cos(lon0_r),  0.0])
    north_hat = np.array([-np.sin(lat0_r)*np.cos(lon0_r),
                           -np.sin(lat0_r)*np.sin(lon0_r), np.cos(lat0_r)])
    up_hat    = np.array([ np.cos(lat0_r)*np.cos(lon0_r),
                            np.cos(lat0_r)*np.sin(lon0_r), np.sin(lat0_r)])
    R_enu = np.vstack([east_hat, north_hat, up_hat])   # (3,3)

    pz = np.array(tbl['z'], dtype=float)
    xyz_ecef_offset = np.column_stack([px, py, pz])    # (N,3) ECEF offsets
    enu_offsets = (R_enu @ xyz_ecef_offset.T).T         # (N,3) ENU offsets

    part_east  = enu_offsets[:, 0] + box_cen[0]
    part_north = enu_offsets[:, 1] + box_cen[1]

    # ── Filter terrain faces within radius ───────────────────────────────────
    face_cen_e = terrain_enu[terrain_faces, 0].mean(axis=1)
    face_cen_n = terrain_enu[terrain_faces, 1].mean(axis=1)
    dist2d = np.hypot(face_cen_e, face_cen_n)
    nearby = np.where(dist2d <= args.radius)[0]
    print(f'  {len(nearby)} terrain faces within {args.radius:.0f} m')

    # Altitude colour limits for visible faces.
    vis_idx = np.unique(terrain_faces[nearby].ravel())
    alt_lo  = terrain_enu[vis_idx, 2].min() if len(vis_idx) else 0
    alt_hi  = terrain_enu[vis_idx, 2].max() if len(vis_idx) else 1000

    # ── Colour scale for kinetic energy ──────────────────────────────────────
    log_ke = np.log10(np.clip(ke, 1e-4, None))
    ke_lo, ke_hi = log_ke.min(), log_ke.max()

    # PDG → marker
    markers = {22: 'o', 11: 's', -11: 's', 13: '^', -13: '^'}
    labels  = {22: 'photon', 11: 'e⁻', -11: 'e⁺', 13: 'μ⁻', -13: 'μ⁺'}

    # ── Figure ────────────────────────────────────────────────────────────────
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(11, 5.5))
    fig.suptitle(
        f'Shower hits — lat={lat0:.3f}° lon={lon0:.3f}°  ({n_part} hits)',
        fontsize=13)

    # ─ Panel 1: 300 m top-down ────────────────────────────────────────────────
    ax1.set_title(f'Top-down  {args.radius:.0f} m view')
    ax1.set_xlabel('East  (m)')
    ax1.set_ylabel('North  (m)')
    ax1.set_aspect('equal')

    # Terrain triangles coloured by altitude.
    if len(nearby):
        triang = mtri.Triangulation(
            terrain_enu[:, 0], terrain_enu[:, 1],
            triangles=terrain_faces[nearby])
        tc = ax1.tripcolor(triang, terrain_enu[:, 2],
                           cmap='terrain', vmin=alt_lo, vmax=alt_hi,
                           shading='flat', alpha=0.8, edgecolors='k', linewidth=0.3)
        fig.colorbar(tc, ax=ax1, label='Altitude  (m)', fraction=0.046, pad=0.04)

    # Box footprint (ENU East vs North).
    e_lo, e_hi = box_enu[:, 0].min(), box_enu[:, 0].max()
    n_lo, n_hi = box_enu[:, 1].min(), box_enu[:, 1].max()
    rect = plt.Polygon([[e_lo,n_lo],[e_hi,n_lo],[e_hi,n_hi],[e_lo,n_hi]],
                       closed=True, fill=False, edgecolor='red', linewidth=2, zorder=3)
    ax1.add_patch(rect)

    # Particle hits.
    cmap_ke = plt.cm.plasma
    if n_part > 0:
        sc = ax1.scatter(part_east, part_north, c=log_ke,
                         cmap='plasma', vmin=ke_lo, vmax=ke_hi,
                         s=40, zorder=4, edgecolors='k', linewidths=0.5)

    ax1.set_xlim(-args.radius, args.radius)
    ax1.set_ylim(-args.radius, args.radius)

    # ─ Panel 2: box zoom ──────────────────────────────────────────────────────
    z = args.zoom
    ax2.set_title(f'Box zoom  ±{z:.0f} m')
    ax2.set_xlabel('East  (m)')
    ax2.set_ylabel('North  (m)')
    ax2.set_aspect('equal')

    # Box triangle edges in ENU (East vs North).
    for face in box_faces:
        idx  = list(face) + [face[0]]
        xs   = box_enu[idx, 0]
        ys   = box_enu[idx, 1]
        ax2.plot(xs, ys, color='red', linewidth=0.8, alpha=0.6)

    # Particle hits.
    if n_part > 0:
        for code in np.unique(pdg):
            mask = pdg == code
            sc2 = ax2.scatter(part_east[mask], part_north[mask],
                              c=log_ke[mask], cmap='plasma',
                              vmin=ke_lo, vmax=ke_hi,
                              s=60, marker=markers.get(code, 'o'),
                              edgecolors='k', linewidths=0.5, zorder=4,
                              label=labels.get(code, str(code)))
        ax2.legend(fontsize=8, loc='upper right')
        cb = fig.colorbar(sc2, ax=ax2, label='log₁₀(KE / GeV)',
                          fraction=0.046, pad=0.04)

    ax2.set_xlim(-z, z)
    ax2.set_ylim(-z, z)

    plt.tight_layout()
    plt.savefig(args.output, dpi=args.dpi, bbox_inches='tight')
    print(f'Saved: {args.output}')

if __name__ == '__main__':
    main()
